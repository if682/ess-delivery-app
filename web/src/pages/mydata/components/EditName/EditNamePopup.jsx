import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';
import { useEffect, useState } from 'react';
import { isInputNull } from '../../../../shared/functions/isInputNull';
import { WarningText } from '../../../../components/atoms/warning-text/WarningText';
import { InputText } from '../../../../components/atoms/input-text/InputText';
import './EditNamePopup.css'

export const EditNamePopup = (props) => {
    const [name, setName] = useState("");
    const [warningMessage, setWarningMessage] = useState(null);
    
    useEffect(() => {
        setWarningMessage(null)
    }, [name])


    async function EditName () {
        if(isInputNull(name)){
            setWarningMessage('Esse campo é de preenchimento obrigatório!')
        }
        else {
            let item = {name: name, email: props.email, password: props.password}

            await fetch(`http://localhost:3001/clients/${props.client_id}`, {
                method: 'PUT',
                headers: {
                'Content-Type': 'application/json'
                },
                body: JSON.stringify(item)
            });
        }
    }

    return (
        <Modal
        {...props}
        size="md"
        aria-labelledby="contained-modal-title-vcenter"
        centered
        >
        <Modal.Header closeButton {...props}>
            <Modal.Title id="contained-modal-title-vcenter">
            Alterar nome
            </Modal.Title>
        </Modal.Header>
        <Modal.Body>
            <div className='input-texts'>
                <InputText setInput={setName} type="text" placeholder="Novo nome"/>
                <WarningText warningMessage={warningMessage}/>
            </div>
            
        </Modal.Body>
        <Modal.Footer>
            <Button className='btn-secondary' onClick={props.onHide}>Cancelar</Button>
            <Button className='btn-primary' onClick={() => EditName()}>Salvar</Button>
        </Modal.Footer>
        </Modal>
    );
}
