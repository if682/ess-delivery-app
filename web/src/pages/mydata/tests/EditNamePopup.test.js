import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { EditNamePopup } from "../components/EditName/EditNamePopup";

describe("EditNamePopup", () => {
    let onHideMock, props;
    
    //mocka a função onHide e o props recebido pelo popup
    beforeEach(() => {
        onHideMock = jest.fn();
        props = {
            email: 'teste@hotmail.com',
            client_id: 123,
            password: 'senha',
            show: true,
            onHide: onHideMock
        };
    });

    //Verifica se o popup é renderizado corretamente
    it('renders correctly', () => {
        const { container } = render(<EditNamePopup {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });
    it('should call onHide when cancel button is clicked', () => {
        const { getByText } = render(<EditNamePopup {...props} />);
        fireEvent.click(getByText('Cancelar'));
        expect(onHideMock).toHaveBeenCalled();
      });

    //Verifica se a mensagem de erro "Esse campo é de preenchimento obrigatório!" é exibida no popup quando se clica em salvar com a caixa de nome vazio
    it('should display warning message "Esse campo é de preenchimento obrigatório!" if either input field is null', async () => {
        const {getByPlaceholderText, getByText} = render(<EditNamePopup {...props} />);
        const nameField = getByPlaceholderText("Novo nome");
        const saveButton = getByText("Salvar");

        fireEvent.change(nameField, { target: { value: "" } });
        fireEvent.click(saveButton);

        const warningMessage = await screen.findByText("Esse campo é de preenchimento obrigatório!");
        expect(warningMessage).toBeInTheDocument();
    });

    //Verifica se a função fetch é chamada para atualizar o nome do usuário quando se clica em salvar com a caixa de texto preenchida
    it("should call the fetch function to update the name when save button is clicked with not null input", async () => {
        global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({ id: 123, name: 'teste2', email: 'teste@hotmail.com', password: 'senha' })
        })
        );
        const { getByText, getByPlaceholderText} = render(<EditNamePopup {...props} />);
        const nameField = getByPlaceholderText("Novo nome");

        fireEvent.change(nameField, { target: { value: "teste2" } });
        fireEvent.click(getByText("Salvar"));
        await waitFor(() => expect(fetch).toHaveBeenCalledWith('http://localhost:3001/clients/123', {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ name: 'teste2', email: 'teste@hotmail.com', password: 'senha'})
        }));

    });
});
