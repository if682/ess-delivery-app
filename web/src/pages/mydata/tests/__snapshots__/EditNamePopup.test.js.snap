// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNamePopup renders correctly 1`] = `null`;
